functions/ts/src/helm_template.ts

```diff
@@ -17,82 +17,214 @@
 import { safeLoadAll } from 'js-yaml';
 import { Configs, isKubernetesObject, generalResult } from 'kpt-functions';
 import { spawnSync } from 'child_process';
+import { tmpdir } from 'os';
+import Path from 'path';
+import { mkdirSync, rmdirSync, readdirSync } from 'fs';
 
 const CHART_NAME = 'name';
-const CHART_PATH = 'chart_path';
-const VALUES = '--values';
+const LOCAL_CHART_PATH = 'local-chart-path';
+const VALUES_PATH = '--values';
+// CHART is chart name that will be processed.
+const CHART = 'chart';
+// CHART_REPO is the repo name that charts are pulled from
+const CHART_REPO = 'chart-repo';
+// CHART_REPO_URL is the repo remote URL
+const CHART_REPO_URL = 'chart-repo-url';
 
-// Render chart templates locally using helm template.
-export async function helmTemplate(configs: Configs) {
-  // Validate config data and read arguments.
-  const args = readArguments(configs);
-  args.unshift('template');
+class HelmTemplateError extends Error {
+  constructor(m: string) {
+    super(m);
 
-  let error;
-  try {
-    const child = spawnSync('helm', args);
-    error = child.stderr;
-    let objects = safeLoadAll(child.stdout);
-    objects = objects.filter((o) => isKubernetesObject(o));
-    configs.insert(...objects);
-  } catch (err) {
-    configs.addResults(generalResult(err, 'error'));
-  }
-  if (error && error.length > 0) {
-    configs.addResults(
-      generalResult(
-        `Helm template command results in error: ${error.toString()}`,
-        'error'
-      )
-    );
+    // Set the prototype explicitly.
+    Object.setPrototypeOf(this, HelmTemplateError.prototype);
   }
 }
 
-function readArguments(configs: Configs) {
-  const args: string[] = [];
-  let nameArg: string | undefined;
-  let pathArg: string | undefined;
-  const configMap = configs.getFunctionConfigMap();
-  if (!configMap) {
-    return args;
+class HelmTemplate {
+  configs: Configs;
+  data: Map<string, string> = new Map<string, string>();
+
+  constructor(configs: Configs) {
+    this.configs = configs;
+
+    this.getData();
   }
-  configMap.forEach((value: string, key: string) => {
-    if (key === CHART_NAME) {
-      nameArg = value;
-    } else if (key === CHART_PATH) {
-      pathArg = value;
-    } else {
+
+  getData() {
+    const data = this.configs.getFunctionConfigMap();
+    if (data !== undefined) {
+      this.data = data;
+    }
+    this.validData();
+  }
+
+  validData() {
```

Comment:
this doesn't need to be a method. In general prefer using first class functions instead of classes and methods. This is a major difference between more functional languages like TS vs something like Java.

This should be:

validateXXX(param)

I'd also avoid use of "data" as it's too generic to be meaningful